tests/js/spec/views/dashboardsV2/gridLayout/dashboard.spec.tsx

```diff
@@ -225,4 +222,43 @@ describe('Dashboards > Dashboard', () => {
       expect(await screen.findByText('Matching Issue Owners Rule')).toBeInTheDocument();
     });
   });
+
+  describe('Edit mode', () => {
+    const widgets = [newWidget];
+    const mount = (dashboard, mockedOrg = initialData.organization) => {
+      const getDashboardComponent = () => (
+        <Dashboard
+          paramDashboardId="1"
+          dashboard={dashboard}
+          organization={mockedOrg}
+          isEditing
+          onUpdate={newWidgets => {
+            widgets.splice(0, widgets.length, ...newWidgets);
+          }}
+          handleUpdateWidgetList={() => undefined}
+          handleAddCustomWidget={() => undefined}
+          onSetWidgetToBeUpdated={() => undefined}
+          router={initialData.router}
+          location={initialData.location}
+          widgetLimitReached={false}
+        />
+      );
+      const {rerender} = rtlMountWithTheme(getDashboardComponent());
+      return {rerender: () => rerender(getDashboardComponent())};
+    };
+
+    it('displays the copy widget button in edit mode', () => {
+      const dashboardWithOneWidget = {...mockDashboard, widgets};
+      mount(dashboardWithOneWidget, organization);
+      expect(screen.getByLabelText('Duplicate Widget')).toBeInTheDocument();
+    });
+
+    it('duplicates the widget', async () => {
+      const dashboardWithOneWidget = {...mockDashboard, widgets};
+      const {rerender} = mount(dashboardWithOneWidget, organization);
+      userEvent.click(screen.getByLabelText('Duplicate Widget'));
+      rerender();
```

Comment:
Do we manually trigger a re-render here because there are a lot of updates that occur after `Duplicate Widget` is clicked? I'm wondering if we've tried wrapping the `userEvent.click` with an `act`

Reply:
We're not testing the `detail.tsx` layer here, just `dashboard.tsx`, which means we need to provide our own logic to update the `dashboard` prop whenever `onUpdate` is called (such as clicking `Duplicate Widget` in this case. This function is typically provided by `detail.tsx`):

```
          onUpdate={newWidgets => {
            widgets.splice(0, widgets.length, ...newWidgets);
          }}
```

As far as I can tell, RTL won't pickup the prop change itself, so we need to call rerender to trigger the update. `act` also doesn't seem to solve this here

Reply:
Ah okay, this is neat. So you're defining `const widgets`, and the `onUpdate` inside `mount` updates that widget array and when we call rerender it uses the new list of widgets as props.

One thing I'd be concerned about is leaking state between tests since that `widgets` variable is shared across tests and is a `const` right? Wondering if we could localize the "widget state" inside the closure of `mount` so it's isolated on every instance of `mount`. e.g.

```js
const mount = dashboard => {
  const widgets = [...dashboard.widgets];
  // everything else
};
```

Reply:
And we would just be overriding the widgets we pass in through the dashboard a bit more explicitly since there's no more global `widgets` variable

Reply:
Good point, I just added a before each to set the `widgets` so there shouldn't be any leaks now. The individual widgets don't get mutated within the tests so I think this should be fine for now!